### ISSUE TITLE
Blossom txns are being received and treated as invalid, causing peer banning and slow sync, before Blossom has activated during IBD

### ISSUE THREAD

**Issue by mdr0id** [zcash/zcash#4283]
Details regarding the reported issue:
https://forum.zcashcommunity.com/t/weird-zcashd-syncing-problem/35694

**Comment by daira** [comment]
I think we've seen a similar issue before with Overwinter and Sapling.

**Comment by rex4539** [comment]
I just had to rsync and I'm seeing hundreds of lines with `Misbehaving: XXXXXXXXX (0 -> 100) BAN THRESHOLD EXCEEDED`

**Comment by LarryRuane** [comment]
I think @daira is right; it's likely that we need an updated version of https://github.com/zcash/zcash/pull/3670. Since I worked on a couple of instances of this problem, I'll see if I can quickly make a PR to fix this.

**Comment by str4d** [comment]
I didn't see this issue (because I was searching for `InitialBlockDownload` and `AcceptToMemoryPool` in error logs, neither of which were in this issue), and opened #4327. Let's use that issue instead, as it has a full write-up.

**Comment by daira** [comment]
@str4d wrote on the duplicate #4327:

I'm syncing one of my mainnet nodes with current master (`v2.1.1-7f02ade06b`), and it was roughly a month behind the chain tip when I started. I noticed that there were periods of time where no blocks were being downloaded because there were no peers. Opening `debug.log`, I find sections like these:

```
2020-02-05 21:40:13 UpdateTip: new best=0000000000ee9ae3705118ba314354e8cdfb19ba5a020fb4bfa9c9e6ec0e82ea  height=610124 bits=469860923 log2_work=56.769573  tx=5347806  date=2019-09-27 02:01:52 progress=0.707112  cache=123.2MiB(14954tx)
2020-02-05 21:40:13 UpdateTip: new best=00000000008d5c0ad6ec9223545e712636146bb7e890604e7c9a82c0835436db  height=610125 bits=469860806 log2_work=56.769582  tx=5347810  date=2019-09-27 02:02:18 progress=0.707113  cache=123.2MiB(14957tx)
2020-02-05 21:40:13 UpdateTip: new best=00000000006377bcede563217f597655920a5ecea95c60bf2da2d3f455c64d16  height=610126 bits=469859423 log2_work=56.769591  tx=5347814  date=2019-09-27 02:03:54 progress=0.707115  cache=123.2MiB(14960tx)
2020-02-05 21:40:13 ERROR: ContextualCheckTransaction(): Sapling spend description invalid
2020-02-05 21:40:13 ERROR: AcceptToMemoryPool: ContextualCheckTransaction failed
2020-02-05 21:40:13 Misbehaving: X.X.X.X:8233 (0 -> 100) BAN THRESHOLD EXCEEDED
```
```
2020-02-05 21:43:36 UpdateTip: new best=0000000000e2d8ecb20c4084f76e010ad56015c7d42efe62d280df7930c3289f  height=615245 bits=469827323 log2_work=56.813807  tx=5379202  date=2019-10-05 23:36:46 progress=0.725202  cache=215.4MiB(22003tx)
2020-02-05 21:43:36 UpdateTip: new best=0000000000eff8aea118f9cb74c0414cd17b14b989a618c2667e92ccf1ef2868  height=615246 bits=469826410 log2_work=56.81382  tx=5379211  date=2019-10-05 23:38:39 progress=0.725205  cache=215.4MiB(22010tx)
2020-02-05 21:43:36 UpdateTip: new best=0000000000b1da429905415f1be07b841163588aa8958dcf64da0b5b9562a68d  height=615247 bits=469824966 log2_work=56.813833  tx=5379216  date=2019-10-05 23:41:37 progress=0.725209  cache=215.4MiB(22014tx)
2020-02-05 21:43:36 ERROR: CScriptCheck(): 2176fb0ecc6297ea0afe281173c4318292975c8d256cb77376c2e28ea22f5585:0 VerifySignature failed: Script evaluated without error but finished with a false/empty top stack element
2020-02-05 21:43:36 ERROR: CScriptCheck(): 2176fb0ecc6297ea0afe281173c4318292975c8d256cb77376c2e28ea22f5585:0 VerifySignature failed: Script evaluated without error but finished with a false/empty top stack element
2020-02-05 21:43:36 ERROR: AcceptToMemoryPool: ConnectInputs failed 2176fb0ecc6297ea0afe281173c4318292975c8d256cb77376c2e28ea22f5585
2020-02-05 21:43:36 Misbehaving: Y.Y.Y.Y:8233 (0 -> 100) BAN THRESHOLD EXCEEDED
```
```
2020-02-05 23:02:03 UpdateTip: new best=0000000000f5a89aad29ec202a0c36b749f2c0e1be95517ea59fb89a2629ad4f  height=649605 bits=469873913 log2_work=57.070786  tx=5586336  date=2019-12-04 22:47:55 progress=0.852771  cache=70.7MiB(11490tx)
2020-02-05 23:02:03 UpdateTip: new best=0000000001155c0cf7a4ba2b5da0c8b51e6949b0a6f420a870db8b85a8bad68b  height=649606 bits=469872971 log2_work=57.070792  tx=5586339  date=2019-12-04 22:48:58 progress=0.852772  cache=70.7MiB(11492tx)
2020-02-05 23:02:03 UpdateTip: new best=00000000004a21b10645b4b8659cabf1b3db41c48bc444d6b77712e58e1122c9  height=649607 bits=469871278 log2_work=57.070798  tx=5586341  date=2019-12-04 22:49:43 progress=0.852774  cache=70.7MiB(11494tx)
2020-02-05 23:02:03 UpdateTip: new best=00000000009e154f9883051166c6c19f857369a4478adddb7dcdf6c6981bda03  height=649608 bits=469869408 log2_work=57.070804  tx=5586346  date=2019-12-04 22:50:49 progress=0.852775  cache=70.7MiB(11499tx)
2020-02-05 23:02:03 ERROR: ContextualCheckTransaction(): Sapling binding signature invalid
2020-02-05 23:02:03 ERROR: AcceptToMemoryPool: ContextualCheckTransaction failed
2020-02-05 23:02:03 Misbehaving: Z.Z.Z.Z:8233 (0 -> 100) BAN THRESHOLD EXCEEDED
```

These all correspond to transaction signature failures (Sapling `spendSig`, transparent, and Sapling `bindingSig` respectively), caused by `AcceptToMemoryPool` trying to validate the transactions using the "currently-active" network upgrade. Unfortunately at this point during IBD, that is Sapling, and so these transactions are rejected - but in doing so, the ban threshold is being set to 100, dropping the peers and preventing reconnects. As we don't have a recently-rejected transactions cache (#2414), this means that a single Blossom-valid transaction can potentially be relayed to us by all connected peers (timing-dependent), causing us to drop them all in quick succession, and then block IBD on being able to connect to new peers.

This previously occurred for the Sprout->Overwinter upgrade in #3399 (for the transaction validity rules introduced then), and was fixed in #3410.

There are several ways in which we could address this:
- Ignore the mempool entirely during IBD, and drop mempool broadcast transactions from peers.
  - This could cause network performance issues during a network upgrade, as peers would not participate in the global mempool for longer than their restart time.
- Don't ban nodes that send us invalid transactions during IBD.
  - This is the stop-gap approach taken in #3410.
- If we are in IBD, our estimated network chain tip height is above some network upgrade activation height, and our current chain tip is below that activation height, then validate transactions against that network upgrade's consensus branch ID instead of what we could currently think it is.
  - This is the fix I suggest in https://github.com/zcash/zcash/pull/3410#issuecomment-408028384 and which @daira disagrees with in https://github.com/zcash/zcash/pull/3410#issuecomment-408070539

Related: #3355, maybe #3642.

**Comment by daira** [comment]
@str4d wrote:
> If we are in IBD, our estimated network chain tip height is above some network upgrade activation height, and our current chain tip is below that activation height, then validate transactions against that network upgrade's consensus branch ID instead of what we could currently think it is.
> * https://github.com/zcash/zcash/pull/3410#issuecomment-408028384 and which @daira disagrees with in https://github.com/zcash/zcash/pull/3410#issuecomment-408070539

To be more specific, what I disagreed with is just estimating the tip height. I don't think we have any need to estimate the tip height in order to achieve the desired effect. If we want to accept transactions from a network upgrade that is in the future relative to the current download height (but known by this release of zcashd) we can do so without estimating the current tip height. This might even be a sensible thing to do *now that we have mempool size limiting* (which we didn't when I originally expressed my disagreement).

**Comment by daira** [comment]
@str4d wrote:
> As we don't have a recently-rejected transactions cache (#2414), this means that a single Blossom-valid transaction can potentially be relayed to us by all connected peers (timing-dependent), causing us to drop them all in quick succession, and then block IBD on being able to connect to new peers.

We do have a recently-rejected transaction cache for transactions that were rejected due to mempool limiting. We don't have one for transactions that were rejected for other reasons (because we assume that those reasons are normally deterministic, which they aren't in this case).

**Comment by daira** [comment]
So, I think we need to consider more carefully why we ever ban peers for sending us "bad" transactions. It's no protection against deliberate DoS. At most it is a protection against accidental DoS, but it seems to me to be more trouble than its worth.

**Comment by daira** [comment]
It seems from the title of #3670 that it would fix the peer-banning issue for Sapling and Blossom, but if you look at the code, it only does so for Overwinter and Sapling.

**Comment by str4d** [comment]
> We do have a recently-rejected transaction cache for transactions that were rejected due to mempool limiting.

We do not reject transactions due to mempool limiting; we accept (ensuring they are valid) and then evict randomly. I chose my wording carefully.

**Comment by zancas** [comment]
The title of #4327 (closed) is _much_ more informative.  If it's an accurate description of the state, can we please make it the title of this issue?

**Comment by zancas** [comment]
> I didn't see this issue (because I was searching for `InitialBlockDownload` and `AcceptToMemoryPool` in error logs, neither of which were in this issue), and opened #4327. ~Let's use that issue instead, as it has a full write-up.~

Can we please add those terms to the issue title, @mdr0id ?

**Comment by zancas** [comment]
> So, I think we need to consider more carefully why we ever ban peers for sending us "bad" transactions. It's no protection against deliberate DoS. At most it is a protection against accidental DoS, but it seems to me to be more trouble than its worth.

Because deliberate DOS-ers can always send "good" transactions?

**Comment by zancas** [comment]
> So, I think we need to consider more carefully why we ever ban peers for sending us "bad" transactions. It's no protection against deliberate DoS. At most it is a protection against accidental DoS, but it seems to me to be more trouble than its worth.

Is there evidence that banning is beneficial in the wild?

Is it the case that nodes running a sufficiently old version of the protocol are banned unless the implementor adjusts their source, and rebuilds?

**Comment by rex4539** [comment]
So I guess this can be closed now?

**Comment by daira** [comment]
This is fixed in v2.1.1-1. Any more general problems with banning should be in a new issue.
